src/local_scheduler/local_scheduler.cc

```diff
@@ -472,9 +471,10 @@ void process_plasma_notification(event_loop *loop,
   uint8_t *notification = read_message_async(loop, client_sock);
   if (!notification) {
     /* The store has closed the socket. */
-    LocalSchedulerState_free(state);
-    LOG_FATAL(
+    kill(getpid(), SIGTERM);
```

Comment:
Nice job tracking this down!

This was a pretty nasty bug, so probably worth documentation exactly what was going on.

The problem is the following, right?

When a Python script finished, Ray would call `cleanup` in `services.py`, which would kill the plasma store. That would cause this code here to get run and the local scheduler would call `LocalSchedulerState_free` and begin to clean up its state (e.g., killing its workers), but then `services.py` would also send a kill signal to the local scheduler, which would then cause `LocalSchedulerState_free` to start running again and then the array of workers was already freed so it was looking at invalid data and would try to call kill on PID 0, which cause problems (e.g., it looked like it caused the driver to die).

It's not completely clear to me that this solves the problem, but I think that you can't interrupt the SIGINT handler with another SIGINT https://www.gnu.org/software/libc/manual/html_node/Signals-in-Handler.html#Signals-in-Handler, but the SIGINT handler may still get run twice.